README.md
# ansible-graylog-modules
Ansible modules for the [Graylog2/graylog2-server](https://github.com/graylog2/graylog2-server) API

A full example playbook can be found in `main.yml`.

### In Progress

* Indices
* Inputs

### Modules

The following modules are available with the corresponding actions:

* graylog_users
  * create
  * update
  * delete
  * list
* graylog_roles
  * create
  * update
  * delete
  * list
* graylog_streams
  * create_stream
  * create_rule
  * update
  * update_rule
  * delete_stream
  * delete_rule
  * list_streams
  * query_streams - query by stream name (ie: to get stream ID)
* graylog_pipelines
  * create
  * create_rule
  * create_connection
  * update
  * update_rule
  * update_connection
  * parse_rule
  * delete
  * delete_rule
  * list
  * query_pipelines - query by pipeline name (ie: to get pipeline ID)
  * query_rules - query by rule name (ie: to get rule ID)
* graylog_index_sets
  * create
  * update
  * delete
  * list
  * query_index_sets - query by index set name (ie: to get index set ID)
* graylog_collector_configurations
  * list_configurations
  * query_collector_configurations
  * update_snippet
* graylog_ldap
  * get
  * update
  * delete
  * test
* graylog_inputs
  * list_inputs
  * delete
  * create_inputs
  * update

### Examples

#### Users

```
- name: Create Graylog user
  graylog_users:
    action: create
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    username: "{{ username }}"
    full_name: "Ansible User"
    password: "{{ password }}"
    email: "test-email@aol.com"
    roles:
      - "ansible_role"

- name: Get Graylog users
  graylog_users:
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
  register: graylog_users

- name: List users
  debug:
    msg: "{{ graylog_users.json }}"
```

#### Roles

```
- name: Create Graylog role
  graylog_roles:
    action: create
    endpoint: "{{ endpoint }}"
    graylog_user: "admin"
    graylog_password: "{{ graylog_password }}"
    name: "ansible_role"
    description: "Ansible test role"
    permissions:
      - "dashboards:read"
    read_only: "true"

- name: Get Graylog roles
  graylog_roles:
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
  register: graylog_roles

- name: List roles
  debug:
    msg: "{{ graylog_roles.json }}"    
```

#### Streams and Stream Rules

```
- name: Create stream
  graylog_streams:
    action: create
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    title: "test_stream"
    description: "Windows and IIS logs"
    matching_type: "AND"
    remove_matches_from_default_stream: False
    rules:
      - {"field":"message","type":1,"value":"test_stream rule","inverted": false,"description":"test_stream rule"}

- name: Get Graylog streams
  graylog_streams:
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
  register: graylog_streams

- name: Get stream from stream name query
  graylog_streams:
    action: query_streams
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    stream_name: "test_stream"
  register: stream

- name:  List single stream by ID
  graylog_streams:
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    stream_id: "{{ stream.json.id }}"

- name: Create stream rule
  graylog_streams:
    action: create_rule
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    stream_id: "{{ stream.json.id }}"
    description: "Windows Security Logs"
    field: "winlogbeat_log_name"
    type: "1"
    value: "Security"
    inverted: False       
```

#### Pipelines, Pipeline Rules, Stream connections

```
- name: Create pipeline rule
  graylog_pipelines:
    action: create_rule
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    title: "test_rule"
    description: "test"
    source: |
      rule "test_rule_domain_threat_intel"
      when
         has_field("dns_query")
      then
         let dns_query_intel = threat_intel_lookup_domain(to_string($message.dns_query), "dns_query");
         set_fields(dns_query_intel);
      end

- name: Create pipeline
  graylog_pipelines:
    action: create
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    title: "test_pipeline"
    source: |
      pipeline "test_pipeline"
      stage 0 match either
      end
    description: "test_pipeline description"

- name: Get pipeline from pipeline name
  graylog_pipelines:
    action: query_pipelines
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    pipeline_name: "test_pipeline"
  register: pipeline

- name: Update pipeline with new rule
  graylog_pipelines:
    action: update
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    pipeline_id: "{{ pipeline.json.id }}"
    description: "test description update"
    source: |
      pipeline "test_pipeline"
      stage 0 match either
      rule "test_rule_domain_threat_intel"
      end

- name: Create Stream connection to processing pipeline
  graylog_pipelines:
    action: create_connection
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    pipeline_id: "{{ pipeline.json.id }}"
    stream_ids:
      - "{{ stream.json.id }}"
```

#### Index Sets and attach Streams

```
- name: Create index set
  graylog_index_sets:
    action: create
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    title: "test_index_set"
    index_prefix: "test_index_"
    description: "test index set"

- name: Get index set by name
  graylog_index_sets:
    action: query_index_sets
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    title: "test_index_set"
  register: index_set

- name: Update stream to use new index set
  graylog_streams:
    action: update
    endpoint: "{{ endpoint }}"
    graylog_user: "{{ graylog_user }}"
    graylog_password: "{{ graylog_password }}"
    stream_id: "{{ stream.json.id }}"
    index_set_id: "{{ index_set.json.id }}"
```

#### LDAP configuration
```
- name: Setup Active Directory authentication without SSL and set "Reader" as default role
  graylog_ldap:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"
    enabled: "true"
    active_directory: "true"
    ldap_uri: "ldap://domaincontroller.mydomain.com:389"
    system_password_set: "true"
    system_username: "ldapbind@mydomain.com"
    system_password: "bindPassw0rd"
    search_base: "cn=users,dc=mydomain,dc=com"
    search_pattern: "(&(objectClass=user)(sAMAccountName={0}))"    
    display_name_attribute: "displayName"
    group_search_base: "cn=groups,dc=mydomain,dc=com"
    group_search_pattern: "(&(objectClass=group)(cn=graylog*))"
    group_id_attribute: "cn"

- name: Remove current LDAP authentication configuration
  graylog_ldap:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"
    action: "delete"

-name: Get current LDAP authentication configuration
  graylog_ldap:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"   
    action: "get"
  register: currentConfiguration

- name: Print current LDAP authentication configuration
  debug:
    msg: "{{ currentConfiguration }}"

- name: Test LDAP bind
  graylog_ldap:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"
    action: "test"
    active_directory: "true"
    ldap_uri: "ldap://domaincontroller.mydomain.com:389"
    system_password_set: "true"
    system_username: "ldapbind@mydomain.com"
    system_password: "bindPassw0rd"
```

#### LDAP group mapping
```
- name: Get all LDAP groups
  graylog_ldap_groups:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"
    action: "list"

- name: Get the LDAP group to Graylog role mapping
  graylog_ldap_groups:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"
    action: "list_mapping"

- name: Update the LDAP group to Graylog role mapping
  graylog_ldap_groups:
    endpoint: "graylog.mydomain.com"
    graylog_user: "username"
    graylog_password: "password"   
    action: "update"
    group: "{{ item.group }}"
    role: "{{ item.role }}"
  with_items:
    - { group : "ldap-group-admins", role : "Admin" }
    - { group : "ldap-group-read", role : "Reader" }
```

#### Input managment
```
  - name: Display all inputs
    graylog_input:
      endpoint: "{{ graylog_endpoint }}"
      graylog_user: "{{ graylog_user }}"
      graylog_password: "{{ graylog_password }}"
      allow_http: "true"
      validate_certs: "false"
      action: "list"

  - name: Remove input with ID 1df0f1234abcd0000d0adf20
    graylog_input:
      endpoint: "{{ graylog_endpoint }}"
      graylog_user: "{{ graylog_user }}"
      graylog_password: "{{ graylog_password }}"
      allow_http: "true"
      validate_certs: "false"
      action: "delete"        
      input_id: "1df0f1234abcd0000d0adf20"

  - name: Create Rsyslog TCP input
    graylog_input_syslog:
      endpoint: "{{ graylog_endpoint }}"
      graylog_user: "{{ graylog_user }}"
      graylog_password: "{{ graylog_password }}"
      allow_http: "true"
      validate_certs: "false"
      action: "create"
      input_type: "TCP"
      title: "Rsyslog TCP"
      global_input: "true"
      allow_override_date: "true"
      bind_address: "0.0.0.0"
      expand_structured_data: "false"
      force_rdns: "false"
      number_worker_threads: "2"
      port: "514"
      recv_buffer_size: "1048576"
      store_full_message: "true"

  - name: Create GELF HTTP input
    graylog_input_gelf:
      endpoint: "{{ graylog_endpoint }}"
      graylog_user: "{{ graylog_user }}"
      graylog_password: "{{ graylog_password }}"
      allow_http: "true"
      validate_certs: "false"
      action: "create"
      input_type: "HTTP"
      title: "Test input GELF HTTP"
      global_input: "true"
      bind_address: "0.0.0.0"

  - name: Update existing RSyslog input
    graylog_input_rsyslog:
      endpoint: "{{ graylog_endpoint }}"
      graylog_user: "{{ graylog_user }}"
      graylog_password: "{{ graylog_password }}"
      allow_http: "true"
      validate_certs: "false"
      action: "update"
      input_type: "TCP"
      title: "Rsyslog TCP"
      global_input: "true"
      allow_override_date: "true"
      expand_structured_data: "false"
      force_rdns: "true"
      port: "1514"
      store_full_message: "true"
      input_id: "1df0f1234abcd0000d0adf20"      
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Released
### Added
- added get_input_id function used when id is not provided for delete action and for creation of extractors
- added create_extractors

### Changed
- module should work with python 2 and 3, so urlparse is used if version is lower than 3, urllib.parse
- delete action on inputs uses name of the input


### Removed

## [1.0.0] 2022-01-27


library/graylog_input.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright: (c) 2019, Matthieu SIMON
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
module: graylog_input
short_description: Manage Graylog inputs
description:
    - The Graylog inputs module allows configuration of inputs nodes.
version_added: "2.9"
author: "Matthieu SIMON"
options:
  graylog_fqdn:
    description:
      - Graylog DNS fqdn. (i.e. graylog.mydomain.com).
    required: false
    type: str    
  graylog_port:
    description:
      - Graylog API port. (i.e. 9000).
    required: false
    type: str
  graylog_user:
    description:
      - Graylog privileged user username.
    required: false
    type: str
  graylog_password:
    description:
      - Graylog privileged user password.
    required: false
    type: str
  allow_http:
    description:
      - Allow non HTTPS connexion
    required: false
    default: false
    type: bool    
  validate_certs:
    description:
      - Allow untrusted certificate
    required: false
    default: false
    type: bool     
  action:
    description:
      - Action to take against LDAP API.
    required: true
    default: list
    choices: [ list, delete ]
    type: str
  input_id:
    description:
      - ID of input to remove
    required: false
    type: str
'''

EXAMPLES = '''
    - name: Display all inputs
      graylog_input:
        endpoint: "{{ graylog_endpoint }}"
        graylog_user: "{{ graylog_user }}"
        graylog_password: "{{ graylog_password }}"
        allow_http: "true"
        validate_certs: "false"
        action: "list"

    - name: Remove input with ID 1df0f1234abcd0000d0adf20
      graylog_input:
        endpoint: "{{ graylog_endpoint }}"
        graylog_user: "{{ graylog_user }}"
        graylog_password: "{{ graylog_password }}"
        allow_http: "true"
        validate_certs: "false"
        action: "delete"        
        input_id: "1df0f1234abcd0000d0adf20"
'''

# import module snippets
import json
import base64
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url, to_text
#from module_utils.graylog_get_api_token import get_token

def delete(module, base_url, headers):

    url = base_url + "/" + module.params['input_id']

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='DELETE')

    if info['status'] != 204:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
    except AttributeError:
        content = info.pop('body', '')

    return info['status'], info['msg'], content, url

def list(module, base_url, headers):

    url = base_url

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')

    if info['status'] != 200:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
        #raise Exception(content)
    except AttributeError:
        content = info.pop('body', '')

    return info['status'], info['msg'], content, url

def get_token(module, endpoint, username, password, allow_http):

    headers = '{ "Content-Type": "application/json", "X-Requested-By": "Graylog API", "Accept": "application/json" }'

    url = endpoint + "/api/system/sessions"

    payload = {}
    payload['username'] = username
    payload['password'] = password
    payload['host'] = endpoint

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))

    if info['status'] != 200:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
        session = json.loads(content)
    except AttributeError:
        content = info.pop('body', '')

    session_string = session['session_id'] + ":session"
    session_bytes = session_string.encode('utf-8')
    session_token = base64.b64encode(session_bytes)

    return session_token

def main():
    module = AnsibleModule(
        argument_spec=dict(
            graylog_fqdn=dict(type='str'),
            graylog_port=dict(type='str'),
            graylog_user=dict(type='str'),
            graylog_password=dict(type='str', no_log=True),
            validate_certs=dict(type='bool', required=False, default=True),
            allow_http=dict(type='bool', required=False, default=False),
            action=dict(type='str', required=False, default='list', 
                        choices=[ 'list' , 'delete' ]),
            input_id=dict(type='str', required=False ),
        )
    )

    graylog_fqdn = module.params['graylog_fqdn']
    graylog_port = module.params['graylog_port']
    graylog_user = module.params['graylog_user']
    graylog_password = module.params['graylog_password']
    action = module.params['action']
    allow_http = module.params['allow_http']

    if allow_http == True:
      endpoint = "http://" + graylog_fqdn + ":" + graylog_port
    else:
      endpoint = "https://" + graylog_fqdn + ":" + graylog_port

    base_url = endpoint + "/api/system/inputs"

    api_token = get_token(module, endpoint, graylog_user, graylog_password, allow_http)
    headers = '{ "Content-Type": "application/json", \
                 "X-Requested-By": "Graylog API", \
                 "Accept": "application/json", \
                 "Authorization": "Basic ' + api_token.decode() + '" }'

    if action == "list":
        status, message, content, url = list(module, base_url, headers)                
    elif action == "delete":
        status, message, content, url = delete(module, base_url, headers)
       
    uresp = {}
    content = to_text(content, encoding='UTF-8')

    try:
        js = json.loads(content)
        #raise Exception(js)
    except ValueError:
        js = ""

    uresp['json'] = js
    uresp['status'] = status
    uresp['msg'] = message
    uresp['url'] = url

    module.exit_json(**uresp)


if __name__ == '__main__':
    main()


library/module_utils/graylog_get_api_token.py
import json
import base64
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url, to_text

#TODO: figure out how to import this into other files to DRY this code out
def get_token(module, endpoint, username, password, allow_http):

    headers = '{ "Content-Type": "application/json", "X-Requested-By": "Graylog API", "Accept": "application/json" }'

    url = endpoint + "/api/system/sessions"

    payload = {}
    payload['username'] = username
    payload['password'] = password
    payload['host'] = endpoint

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))

    if info['status'] != 200:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
        session = json.loads(content)
    except AttributeError:
        content = info.pop('body', '')

    session_string = session['session_id'] + ":session"
    session_bytes = session_string.encode('utf-8')
    session_token = base64.b64encode(session_bytes)

    return session_token


library/graylog_input_gelf.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright: (c) 2019, Matthieu SIMON
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
module: graylog_inputs
short_description: Manage Graylog inputs
description:
    - The Graylog inputs module allows configuration of inputs nodes.
version_added: "2.9"
author: "Matthieu SIMON"
options:
  endpoint:
    description:
      - Graylog endoint. (i.e. graylog.mydomain.com).
    required: false
    type: str
  graylog_user:
    description:
      - Graylog privileged user username.
    required: false
    type: str
  graylog_password:
    description:
      - Graylog privileged user password.
    required: false
    type: str
  allow_http:
    description:
      - Allow non HTTPS connexion
    required: false
    default: false
    type: bool
  validate_certs:
    description:
      - Allow untrusted certificate
    required: false
    default: false
    type: bool
  action:
    description:
      - Action to take against system/input API.
      - Warning : when update, all settings with default value set in this Ansible module (like bind_address, port ...) will replace existing values
        You must explicitly set these values if they differ from those by default
    required: true
    default: create
    choices: [ create, update ]
    type: str
  input_type:
    description:
      - Input type (not all are implemented at this time)
    required: true
    default: UDP
    choices: [ 'UDP', 'TCP', 'HTTP' ]
    type: str
  title:
    description:
      - Entitled of the input
      - Required with actions create, update and delete
    required: true
    type: str
  input_id:
    description:
      - ID of input to update
    required: false
    type: str
  global_input:
    description:
      - Input is present on all Graylog nodes
    required: false
    default: true
    type: bool
  node:
    description:
      - Node name if input is not global
    required: false
    type: str
  bind_address:
    description:
      - Address to listen on
    required: false
    default: "0.0.0.0"
    type: str
  port:
    description:
      - Port to listen on
    required: true
    default: 12201
    type: int
   number_worker_threads:
    description:
      - Number of worker threads processing network connections for this input.
    required: false
    default: 2
    type: int
  override_source:
    description:
      - The source is a hostname derived from the received packet by default. Set this if you want to override it with a custom string.
    required: false
    type: str
  recv_buffer_size:
    description:
      - The size in bytes of the recvBufferSize for network connections to this input.
    required: false
    default: 1048576
    type: int
  tcp_keepalive:
    description:
      - Enable TCP keepalive packets (TCP & HTTP only)
    required: false
    default: false
    type: bool
  tls_enable:
    description:
      - Accept TLS connections (TCP & HTTP only)
    required: false
    default: false
    type: bool
  tls_cert_file:
    description:
      - Path to the TLS certificate file (TCP & HTTP only)
    required: false
    type: str
  tls_key_file:
    description:
      - Path to the TLS private key file (TCP & HTTP only)
    required: false
    type: str
  tls_key_password:
    description:
      - The password for the encrypted key file. (TCP & HTTP only)
    required: false
    type: str
  tls_client_auth:
    description:
      - Whether clients need to authenticate themselves in a TLS connection (TCP & HTTP only)
    required: false
    default: disabled
    choices: [ 'disabled', 'optional', 'required' ]
  tls_client_auth_cert_file:
    description:
      - TLS Client Auth Trusted Certs (File or Directory) (TCP & HTTP only)
    required: false
    type: str
  use_null_delimiter:
    description:
      - Use null byte as frame delimiter ? Otherwise newline delimiter is used. (TCP Only)
    required: false
    default: false
    type: bool
  decompress_size_limit:
    description:
      - The maximum number of bytes after decompression.
    required: false
    default: 8388608
    type: int
  enable_cors:
    description:
      - Input sends CORS headers to satisfy browser security policies (HTTP Only)
    required: false
    default: true
    type: bool
  idle_writer_timeout:
    description:
      - The server closes the connection after the given time in seconds after the last client write request. (use 0 to disable) (HTTP Only)
    required: false
    default: 60
    type: int
  max_chunk_size:
    description:
      - The maximum HTTP chunk size in bytes (e. g. length of HTTP request body) (HTTP Only)
    required: false
    default: 65536
    type: int
  max_message_size:
    description:
      - The maximum length of a message. (TCP Only)
    required: false
    default: 2097152
    type: int
'''

EXAMPLES = '''
    - name: Create GELF HTTP input
      graylog_input_gelf:
        endpoint: "{{ graylog_endpoint }}"
        graylog_user: "{{ graylog_user }}"
        graylog_password: "{{ graylog_password }}"
        allow_http: "true"
        validate_certs: "false"
        action: "create"
        input_type: "HTTP"
        title: "Test input GELF HTTP"
        global_input: "true"
        bind_address: "0.0.0.0"
'''

# import module snippets
import json
import base64
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url, to_text
import re

def query_inputs(module, base_url, headers, title):

    url = base_url
    inputExist = False

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')

    if info['status'] != 200:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
        data = json.loads(content)
    except AttributeError:
        content = info.pop('body', '')

    regex = r"^" + re.escape(title) + r"$"

    for graylogInputs in data['inputs']:
      if re.match(regex, graylogInputs['title']) is not None:
        inputExist = True

    return inputExist


def create_input(module, base_url, headers):
    """
     Check if an input with same title exists, if so update/PUT it, if not create/POST it.
    :param module: Ansible playbook parameters
    :type module: str
    :param base_url: Graylog API URL
    :type base_url: str
    :param headers: HTTP Headers to be sent to Graylog API
    :type headers: dict
    :return: HTTP Response Code, message, response body, and API URL
    :rtype: tuple
    """

    url = base_url

    if module.params['action'] == "create":
      inputExist = query_inputs(module, base_url, headers, module.params['title'])
      if inputExist == True:
        #TODO: use this exit_json in streams module also!
        module.exit_json(changed=False)
      httpMethod = "POST"
    else:
      httpMethod = "PUT"
      url = base_url + "/" + module.params['input_id']

    configuration = {}
   #TODO: add conditionals here based on new module param log_format to accomodate all input types, no need to created separate functions for each afaik
    for key in [ 'bind_address', 'port', 'number_worker_threads', 'override_source', 'recv_buffer_size', \
                 'tcp_keepalive', 'tls_enable', 'tls_cert_file', 'tls_key_file', 'tls_key_password', \
                 'tls_client_auth', 'tls_client_auth_cert_file', 'use_null_delimiter', 'decompress_size_limit', \
                 'enable_cors', 'idle_writer_timeout', 'max_chunk_size', 'max_message_size' ]:
        if module.params[key] is not None:
            configuration[key] = module.params[key]

    payload = {}

    payload['type'] = module.params['input_type']
    payload['title'] = module.params['title']
    payload['global'] = module.params['global_input']
    payload['node'] = module.params['node']
    payload['configuration'] = configuration

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method=httpMethod, data=module.jsonify(payload))

    if info['status'] != 201:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
    except AttributeError:
        content = info.pop('body', '')

    return info['status'], info['msg'], content, base_url

def get_token(module, endpoint, username, password, allow_http):

    headers = '{ "Content-Type": "application/json", "X-Requested-By": "Graylog API", "Accept": "application/json" }'

    url = endpoint + "/api/system/sessions"

    payload = {}
    payload['username'] = username
    payload['password'] = password
    payload['host'] = endpoint

    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))

    if info['status'] != 200:
        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))

    try:
        content = to_text(response.read(), errors='surrogate_or_strict')
        session = json.loads(content)
    except AttributeError:
        content = info.pop('body', '')

    session_string = session['session_id'] + ":session"
    session_bytes = session_string.encode('utf-8')
    session_token = base64.b64encode(session_bytes)

    return session_token

def main():
    module = AnsibleModule(
        argument_spec=dict(
            graylog_fqdn=dict(type='str'),
            graylog_port=dict(type='str'),
            graylog_user=dict(type='str'),
            graylog_password=dict(type='str', no_log=True),
            validate_certs=dict(type='bool', required=False, default=True),
            allow_http=dict(type='bool', required=False, default=False),
            action=dict(type='str', required=False, default='create',
                        choices=[ 'create', 'update' ]),
            input_type=dict(type='str', required=False, default='UDP',
                        choices=[ 'UDP', 'TCP', 'HTTP' ]),
            title=dict(type='str', required=True ),
            global_input=dict(type='bool', required=False, default=True),
            node=dict(type='str', required=False),
            bind_address=dict(type='str', required=False, default='0.0.0.0'),
            port=dict(type='int', required=False, default=12201),
            number_worker_threads=dict(type='int', required=False, default=2),
            override_source=dict(type='str', required=False),
            recv_buffer_size=dict(type='int', required=False, default=1048576),
            tcp_keepalive=dict(type='bool', required=False, default=False),
            tls_enable=dict(type='bool', required=False, default=False),
            tls_cert_file=dict(type='str', required=False),
            tls_key_file=dict(type='str', required=False),
            tls_key_password=dict(type='str', required=False, no_log=True),
            tls_client_auth=dict(type='str', required=False, default='disabled',
                        choices=[ 'disabled', 'optional', 'required' ]),
            tls_client_auth_cert_file=dict(type='str', required=False),
            use_null_delimiter=dict(type='bool', required=False, default=False),
            decompress_size_limit=dict(type='int', required=False, default=8388608),
            enable_cors=dict(type='bool', required=False, default=True),
            idle_writer_timeout=dict(type='int', required=False, default=60),
            max_chunk_size=dict(type='int', required=False, default=65536),
            max_message_size=dict(type='int', required=False, default=2097152)
        )
    )

    graylog_fqdn = module.params['graylog_fqdn']
    graylog_port = module.params['graylog_port']
    graylog_user = module.params['graylog_user']
    graylog_password = module.params['graylog_password']
    allow_http = module.params['allow_http']

    if allow_http == True:
      endpoint = "http://" + graylog_fqdn + ":" + graylog_port
    else:
      endpoint = "https://" + graylog_fqdn + ":" + graylog_port

    # Build full name of input type
    if module.params['input_type'] == "TCP":
        module.params['input_type'] = "org.graylog2.inputs.gelf.tcp.GELFTCPInput"
    elif module.params['input_type'] == "UDP":
        module.params['input_type'] = "org.graylog2.inputs.gelf.udp.GELFUDPInput"
    else:
        module.params['input_type'] = "org.graylog2.inputs.gelf.http.GELFHttpInput"

    base_url = endpoint + "/api/system/inputs"

    api_token = get_token(module, endpoint, graylog_user, graylog_password, allow_http)
    headers = '{ "Content-Type": "application/json", \
                 "X-Requested-By": "Graylog API", \
                 "Accept": "application/json", \
                 "Authorization": "Basic ' + api_token.decode() + '" }'

    # create the input if one with same title does not exist
    status, message, content, url = create_input(module, base_url, headers)

    uresp = {}
    content = to_text(content, encoding='UTF-8')

    try:
        js = json.loads(content)
    except ValueError:
        js = ""

    uresp['json'] = js
    uresp['status'] = status
    uresp['msg'] = message
    uresp['url'] = url

    module.exit_json(**uresp)


if __name__ == '__main__':
    main()
